### PR TITLE
Implement WebSocket support based on Plug.Conn.Adapter.upgrade/3

### PR DESCRIPTION
**Pull request by mtrudel**
This PR provides support for upgrading HTTP connections to WebSocket connections by use of the `c:Plug.Conn.Adapter.upgrade/3` callback. WebSocket implementations are called based on the Sock API (which mirrors `Phoenix.Socket.Transport`) . This work is the Bandit analog of https://github.com/elixir-plug/plug/pull/1119

See https://github.com/phoenixframework/phoenix/issues/5003 for an omnibus overview of the larger body of work that this PR is part of.

### CODE REVIEW

**Comment by mtrudel** [comment]
@moogle19 - would you be able to take a look at this in the context of https://github.com/phoenixframework/phoenix/issues/5003?

File: lib/bandit/websocket/socket.ex [hunk 1]
```diff
@@ -1,56 +1,48 @@
 defprotocol Bandit.WebSocket.Socket do
   @moduledoc false
   #
-  # A protocol adding close behaviour to the Sock.Socket protocol
+  # A protocol adding defining the low-level functionality of a WebSocket
```
Comment (moogle19) [inline lib/bandit/websocket/socket.ex:4]:
```suggestion
  # A protocol defining the low-level functionality of a WebSocket
```

File: lib/bandit/websocket/handler.ex [hunk 2]
```diff
@@ -1,31 +1,32 @@
...
+          timeout -> {:continue, Map.put(state, :connection, connection), {:persistent, timeout}}
+        end
 
-    Connection.handle_connection(state.conn, socket, connection)
-    |> case do
-      {:continue, connection} -> {:continue, update_state(state, connection)}
-      {:continue, connection, timeout} -> {:continue, update_state(state, connection), timeout}
-      {:close, _} -> {:close, state}
+      {:error, reason, connection} ->
+        {:error, reason, state |> Map.put(:connection, connection)}
```
Comment (moogle19) [inline lib/bandit/websocket/handler.ex:26]:
```suggestion
        {:error, reason, Map.put(state, :connection, connection)}
```
to be consistent with the previous cases

File: lib/bandit/websocket/connection.ex [hunk 3]
```diff
@@ -53,7 +29,7 @@ defmodule Bandit.WebSocket.Connection do
 
       %Frame.Text{fin: true} = frame ->
         if String.valid?(frame.data) do
-          connection.sock.handle_text_frame(frame.data, socket, connection.sock_state)
+          connection.sock.handle_in({frame.data, opcode: :text}, connection.sock_state)
```
Comment (moogle19) [inline lib/bandit/websocket/connection.ex:32]: More of a `Sock` issue, but what is the intention of using a keyword list as second tuple argument and not just the opcode?

**Review by moogle19 — COMMENTED** [review]
Apart from a few cosmetic changes this looks awesome! 🙂

File: lib/bandit/websocket/connection.ex (hunk 3 above)
Comment (mtrudel) [reply lib/bandit/websocket/connection.ex:32]: It's a concession to parity with the existing `Phoenix.Socket.Transport` behaviour (specifically, https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/socket/transport.ex#L163-L200). I agree that it's a bit awkward (especially as it doesn't have parity with the shape of return values in `{:push, ...` and `{:reply, ...` tuples), but it allows Phoenix to serve `Phoenix.Socket` instances as valid, conforming Sock instances.

**Comment by mtrudel** [comment]
Thanks for the review @moogle19 !

**Pull request merged**